examples/async-sentiment.rs

```diff
@@ -0,0 +1,63 @@
+use std::{
+    sync::mpsc,
+    thread::{self, JoinHandle},
+};
+
+use anyhow::Result;
+use rust_bert::pipelines::sentiment::{Sentiment, SentimentConfig, SentimentModel};
+use tokio::{sync::oneshot, task};
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let (_handle, classifier) = SentimentClassifier::spawn();
+
+    let texts = vec![
+        "Classify this positive text".to_owned(),
+        "Classify this negative text".to_owned(),
+    ];
+    let sentiments = classifier.predict(texts).await?;
+    println!("Results: {:?}", sentiments);
+
+    Ok(())
+}
+
+/// Message type for internal channel, passing around texts and return value
+/// senders
+type Message = (Vec<String>, oneshot::Sender<Vec<Sentiment>>);
+
+/// Runner for sentiment classification
+#[derive(Debug, Clone)]
+pub struct SentimentClassifier {
+    sender: mpsc::SyncSender<Message>,
+}
+
+impl SentimentClassifier {
+    /// Spawn a classifier on a separate thread and return a classifier instance
+    /// to interact with it
+    pub fn spawn() -> (JoinHandle<Result<()>>, SentimentClassifier) {
+        let (sender, receiver) = mpsc::sync_channel(100);
+        let handle = thread::spawn(move || Self::runner(receiver));
+        (handle, SentimentClassifier { sender })
+    }
+
+    /// The classification runner itself
+    fn runner(receiver: mpsc::Receiver<Message>) -> Result<()> {
```

Comment:
I am not an expert with Tokio and async runtimes - could you please share what is happening here? Creating models is a heavy process, ideally a model is created once and can then serve multiple request. Note also that models are thread safe and can therefore be created once and then shared between multiple threads. 

The only "blocking" part is the model instantiation when **remote resources** are used. Using local resources should not block any Tokio runtime and should therefore be compatible in fully async workloads. It looks like the message here are sent synchronously - is this required?

Reply:
Unfortunately, models do not seem to be thread-safe at all. As far as I can see, they are implicitly implementing `!Send` & `!Sync` to avoid asynchronous sharing, which means it can only be shared inside `Arc<Mutex<...>>`, not even `RwLock` sadly. But again, using a lock and executing the prediction inside a future is not recommended due to extensive CPU usage over a longer time without yielding.

This runner is executed on a newly created OS thread, outside of any async runtime, and simply executes all prediction requests:
1. Accepts incoming prediction requests via the multi-producer, single-consumer channel (sync, because receiving in sync code)
2. Running the model's prediction on it
3. Sending the result via the oneshot channel back

The async oneshot channel allows the handler to run the prediction function in an async runtime without anything to take care of, as it can be simply `.await`ed.

The model itself will only be created once though, as long as the spawn function is only called once. The returned handler to interact with the runner and predict samples can easily be shared and cloned without duplicating the runner or the model.

Reply:
I have to correct myself, it doesn't seem to be implicit and it is `Send` indeed. However, it is not `Sync` and there cannot be shared without a mutex across threads (or across a non-local async runtime).

Reply:
Thank you for the clarifications and for the pull request!